File: 백준/Gold/1891. 사분면/사분면.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int d = Integer.parseInt(st.nextToken());
        String num = st.nextToken();

        st = new StringTokenizer(br.readLine());

        long x = Long.parseLong(st.nextToken());
        long y = Long.parseLong(st.nextToken());

        long size = (long) Math.pow(2, d - 1);
        long[] pos = findPos(0, 0, size, num, 0);

        long nx = pos[0] - y;
        long ny = pos[1] + x;

        if (nx < 0 || ny < 0 || nx >= size * 2L || ny >= size * 2L) {
            System.out.println(-1);
            return;
        }

        System.out.println(findAns(nx, ny, size, d));
    }

    private static long[] findPos(long x, long y, long size, String num, int index) {
        if (index == num.length()) {
            return new long[]{x, y};
        }

        char c = num.charAt(index);
        long half = size / 2;

        switch (c) {
            case '1': return findPos(x, y + size, half, num, index + 1);
            case '2': return findPos(x, y, half, num, index + 1);
            case '3': return findPos(x + size, y, half, num, index + 1);
            case '4': return findPos(x + size, y + size, half, num, index + 1);
        }

        return new long[]{};
    }

    private static String findAns(long x, long y, long size, int d) {
        if (d == 0) {
            return "";
        }

        long half = size / 2;

        if (x < size && y >= size) return "1" + findAns(x, y - size, half, d - 1);
        else if (x < size && y < size) return "2" + findAns(x, y, half, d - 1);
        else if (x >= size && y < size) return "3" + findAns(x - size, y, half, d - 1);
        else return "4" + findAns(x - size, y - size, half, d - 1);
    }
}

File: 백준/Gold/1891. 사분면/사분면.py
def find_pos(x, y, size, num, index):
    if index == len(num):
        return x, y

    c = num[index]
    half = size // 2

    if c == '1':
        return find_pos(x, y + size, half, num, index + 1)
    elif c == '2':
        return find_pos(x, y, half, num, index + 1)
    elif c == '3':
        return find_pos(x + size, y, half, num, index + 1)
    else:
        return find_pos(x + size, y + size, half, num, index + 1)


def find_ans(x, y, size, d):
    if d == 0:
        return ""

    half = size // 2

    if x < size <= y:
        return "1" + find_ans(x, y - size, half, d - 1)
    elif x < size and y < size:
        return "2" + find_ans(x, y, half, d - 1)
    elif x >= size > y:
        return "3" + find_ans(x - size, y, half, d - 1)
    else:
        return "4" + find_ans(x - size, y - size, half, d - 1)


data = input().split()
d = int(data[0])
num = data[1]

x, y = map(int, input().split())

size = 1 << (d - 1)
pos = find_pos(0, 0, size, num, 0)

nx = pos[0] - y
ny = pos[1] + x

if nx < 0 or ny < 0 or nx >= size * 2 or ny >= size * 2:
    print(-1)
else:
    print(find_ans(nx, ny, size, d))
